# Oakland 311 calls

This notebook is part of Assignment #6.

In [6]:
import pandas as pd
import altair as alt
import requests

/Users/derekschwabe/.pyenv/versions/3.8.5/envs/lecture_06-3.8.5/lib/python3.8/site-packages/pandas/compat/__init__.py:124: UserWarning: Could not import the lzma module. Your installed Python is incomplete. Attempting to use lzma compression will result in a RuntimeError.
  warnings.warn(msg)


## Step 1: Downloading the data

Here I am downloading the[311 call data](https://data.oaklandca.gov/Infrastructure/Service-requests-received-by-the-Oakland-Call-Cent/quth-gb8e) from the City of Oakland's Open Data Portal. 

In [2]:
# url = 'https://data.oaklandca.gov/api/views/quth-gb8e/rows.csv?accessType=DOWNLOAD'
# r = requests.get(url, allow_redirects=False)

# # write the content of the request into a file called `oakland_311.csv`
# open('oakland_311.csv', 'wb').write(r.content)

179882250

## Step 2: Importing the Oakland 311 CSV into a Pandas Dataframe

I'm using the same method we used on the Berkeley 311 data, called `pd.read_csv()` to import the Oakland 311 csv file into a dataframe.

I'm making sure to assign the dataframe into a variable. I'll name the dataframe `oakland_311_original`. I will continue using df to refer generically to a dataframe, as we did in lecture. 

In [7]:
oakland_311_original = pd.read_csv('oakland_311.csv')

## Step 3: Viewing the data
Here I am using the df.head() df.and df.tail() to view the first and last five rows of the Oakland 311 dataframe. The purpose of this step is just to familiarize myself with the data structure by taking a first glance. This is especially important for seeing how columns are named, compared to what we saw in the Berkeley 311 dataset. 

In [8]:
oakland_311_original.head()

REQUESTID            DATETIMEINIT       SOURCE  \
0    1145885  09/09/2021 06:08:56 PM  SeeClickFix   
1    1146984  09/14/2021 08:08:50 AM        Phone   
2    1150718  09/27/2021 07:52:03 AM  SeeClickFix   
3    1151336  09/28/2021 03:37:51 PM  SeeClickFix   
4    1151681  09/29/2021 04:18:33 PM  SeeClickFix   

                                  DESCRIPTION  REQCATEGORY  \
0     Illegal Dumping (Enforcement Potential)  ENVIRON_ENF   
1     Illegal Dumping (Enforcement Potential)  ENVIRON_ENF   
2  Illegal Dumping - debris, appliances, etc.      ILLDUMP   
3                 Parking - Abandoned Vehicle       POLICE   
4                 Parking - Abandoned Vehicle       POLICE   

                                  REQADDRESS STATUS REFERREDTO DATETIMECLOSED  \
0  (37.788936670596335, -122.24357499716558)   OPEN        NaN            NaN   
1   (37.78829957727266, -122.23117705915959)   OPEN        NaN            NaN   
2   (37.82142504165362, -122.28732679694832)   OPEN        NaN            NaN   
3   (37.75306101818787, -122.16312297802386)   OPEN        NaN            NaN   
4     (37.785503590812, -122.20544605709021)   OPEN        NaN            NaN   

           SRX          SRY COUNCILDISTRICT BEAT         PROBADDRESS     City  \
0  6057859.838  2114424.905            CCD2  19X       1438 16TH AVE  Oakland   
1  6061437.527  2114126.459            CCD5  21X      2324 E 21ST ST  Oakland   
2  6045445.169  2126491.007            CCD3  07X  ETTIE ST & 28TH ST  Oakland   
3  6080871.846  2100940.775            CCD7  35X    9227 HILLSIDE ST  Oakland   
4  6068853.201  2112972.033            CCD4  24Y  3901 ALLENDALE AVE  Oakland   

  State  
0    CA  
1    CA  
2    CA  
3    CA  
4    CA

In [9]:
oakland_311_original.tail()

REQUESTID            DATETIMEINIT       SOURCE  \
852536    1151935  09/30/2021 01:54:23 PM  SeeClickFix   
852537    1151856  09/30/2021 11:26:13 AM  SeeClickFix   
852538    1147185  09/14/2021 02:54:49 PM  SeeClickFix   
852539    1150481  09/25/2021 03:31:40 PM  SeeClickFix   
852540    1151347  09/28/2021 03:54:42 PM  SeeClickFix   

                                      DESCRIPTION     REQCATEGORY  \
852536          Utility - Lid - Missing or Broken  ROW_INSPECTORS   
852537  Litter - Street Litter Container - Broken         ILLDUMP   
852538                Parking - Abandoned Vehicle          POLICE   
852539    Homeless Encampment - Living in Vehicle        HE_CLEAN   
852540    Homeless Encampment - Living in Vehicle        HE_CLEAN   

                                       REQADDRESS    STATUS REFERREDTO  \
852536    (37.80086333056584, -122.2728949987278)      OPEN        NaN   
852537     (37.7990150403045, -122.2729668018932)      OPEN        NaN   
852538   (37.78429373521214, -122.19783353979535)      OPEN        NaN   
852539   (37.75920504225326, -122.18706682059627)  UNFUNDED        NaN   
852540  (37.781792932032374, -122.24281520941372)  UNFUNDED        NaN   

       DATETIMECLOSED          SRX          SRY COUNCILDISTRICT BEAT  \
852536            NaN  6049471.080  2118926.060            CCD2  03X   
852537            NaN  6049437.589  2118253.569            CCD2  03X   
852538            NaN  6071044.715  2112491.573            CCD6  28X   
852539            NaN  6073990.955  2103301.471            CCD7  30X   
852540            NaN  6058030.915  2111820.075            CCD5  19X   

                          PROBADDRESS     City State  
852536                   988 BROADWAY  Oakland    CA  
852537           7TH ST & FRANKLIN ST  Oakland    CA  
852538             3216 COURTLAND AVE  Oakland    CA  
852539  73RD AVE & INTERNATIONAL BLVD  Oakland    CA  
852540               2000 EMBARCADERO  Oakland    CA

### Specifying the number of first or last rows I want to see
If I want to see more rows than the standard 5, I could specify that in the parentheses after df.head() or df.tail(). Here's an example using df.tail() to look at the last 10 rows of the Oakland 311 data:

In [10]:
oakland_311_original.tail(10)

REQUESTID            DATETIMEINIT       SOURCE  \
852531    1145831  09/09/2021 02:29:54 PM  SeeClickFix   
852532    1146284  09/11/2021 01:51:18 PM  SeeClickFix   
852533    1148083  09/17/2021 09:28:45 AM  SeeClickFix   
852534    1149148  09/22/2021 01:01:32 PM  SeeClickFix   
852535    1145386  09/07/2021 09:43:37 PM  SeeClickFix   
852536    1151935  09/30/2021 01:54:23 PM  SeeClickFix   
852537    1151856  09/30/2021 11:26:13 AM  SeeClickFix   
852538    1147185  09/14/2021 02:54:49 PM  SeeClickFix   
852539    1150481  09/25/2021 03:31:40 PM  SeeClickFix   
852540    1151347  09/28/2021 03:54:42 PM  SeeClickFix   

                                       DESCRIPTION     REQCATEGORY  \
852531  Illegal Dumping - debris, appliances, etc.         ILLDUMP   
852532  Illegal Dumping - debris, appliances, etc.         ILLDUMP   
852533  Illegal Dumping - debris, appliances, etc.         ILLDUMP   
852534        Illegal Dumping � mattress/boxspring         ILLDUMP   
852535  Illegal Dumping - debris, appliances, etc.         ILLDUMP   
852536           Utility - Lid - Missing or Broken  ROW_INSPECTORS   
852537   Litter - Street Litter Container - Broken         ILLDUMP   
852538                 Parking - Abandoned Vehicle          POLICE   
852539     Homeless Encampment - Living in Vehicle        HE_CLEAN   
852540     Homeless Encampment - Living in Vehicle        HE_CLEAN   

                                       REQADDRESS    STATUS REFERREDTO  \
852531   (37.75938430139389, -122.18785989912476)    CLOSED        NaN   
852532   (37.74563358053681, -122.17982370769354)    CLOSED        NaN   
852533  (37.763113280808824, -122.21500621132596)    CLOSED        NaN   
852534  (37.806640777743276, -122.23953621821266)    CLOSED        NaN   
852535   (37.79708238688122, -122.25077991854145)    CLOSED        NaN   
852536    (37.80086333056584, -122.2728949987278)      OPEN        NaN   
852537     (37.7990150403045, -122.2729668018932)      OPEN        NaN   
852538   (37.78429373521214, -122.19783353979535)      OPEN        NaN   
852539   (37.75920504225326, -122.18706682059627)  UNFUNDED        NaN   
852540  (37.781792932032374, -122.24281520941372)  UNFUNDED        NaN   

                DATETIMECLOSED          SRX          SRY COUNCILDISTRICT BEAT  \
852531  09/14/2021 04:39:08 PM  6073762.904  2103370.866            CCD7  26Y   
852532  09/15/2021 08:10:45 PM  6075995.780  2098322.966            CCD7  33X   
852533  09/21/2021 11:05:37 AM  6065941.567  2104871.100            CCD5  26X   
852534  09/27/2021 03:54:19 PM  6059146.529  2120848.542            CCD2  15X   
852535  09/10/2021 07:57:22 PM  6055833.700  2117429.291            CCD2  17X   
852536                     NaN  6049471.080  2118926.060            CCD2  03X   
852537                     NaN  6049437.589  2118253.569            CCD2  03X   
852538                     NaN  6071044.715  2112491.573            CCD6  28X   
852539                     NaN  6073990.955  2103301.471            CCD7  30X   
852540                     NaN  6058030.915  2111820.075            CCD5  19X   

                          PROBADDRESS     City State  
852531        7211 INTERNATIONAL BLVD  Oakland    CA  
852532                      9198 E ST  Oakland    CA  
852533                   601 50TH AVE  Oakland    CA  
852534             845 MACARTHUR BLVD  Oakland    CA  
852535                  537 E 17TH ST  Oakland    CA  
852536                   988 BROADWAY  Oakland    CA  
852537           7TH ST & FRANKLIN ST  Oakland    CA  
852538             3216 COURTLAND AVE  Oakland    CA  
852539  73RD AVE & INTERNATIONAL BLVD  Oakland    CA  
852540               2000 EMBARCADERO  Oakland    CA

### Calling the dataframe to view both the first and last five rows
I can also call the dataframe on its own to see both the first 5 rows and the last 5 rows in one command. The elipses in between the first and last 5 rows indicate that all the remaining rows are in between. Here's what that looks like:

In [11]:
oakland_311_original

REQUESTID            DATETIMEINIT       SOURCE  \
0         1145885  09/09/2021 06:08:56 PM  SeeClickFix   
1         1146984  09/14/2021 08:08:50 AM        Phone   
2         1150718  09/27/2021 07:52:03 AM  SeeClickFix   
3         1151336  09/28/2021 03:37:51 PM  SeeClickFix   
4         1151681  09/29/2021 04:18:33 PM  SeeClickFix   
...           ...                     ...          ...   
852536    1151935  09/30/2021 01:54:23 PM  SeeClickFix   
852537    1151856  09/30/2021 11:26:13 AM  SeeClickFix   
852538    1147185  09/14/2021 02:54:49 PM  SeeClickFix   
852539    1150481  09/25/2021 03:31:40 PM  SeeClickFix   
852540    1151347  09/28/2021 03:54:42 PM  SeeClickFix   

                                       DESCRIPTION     REQCATEGORY  \
0          Illegal Dumping (Enforcement Potential)     ENVIRON_ENF   
1          Illegal Dumping (Enforcement Potential)     ENVIRON_ENF   
2       Illegal Dumping - debris, appliances, etc.         ILLDUMP   
3                      Parking - Abandoned Vehicle          POLICE   
4                      Parking - Abandoned Vehicle          POLICE   
...                                            ...             ...   
852536           Utility - Lid - Missing or Broken  ROW_INSPECTORS   
852537   Litter - Street Litter Container - Broken         ILLDUMP   
852538                 Parking - Abandoned Vehicle          POLICE   
852539     Homeless Encampment - Living in Vehicle        HE_CLEAN   
852540     Homeless Encampment - Living in Vehicle        HE_CLEAN   

                                       REQADDRESS    STATUS REFERREDTO  \
0       (37.788936670596335, -122.24357499716558)      OPEN        NaN   
1        (37.78829957727266, -122.23117705915959)      OPEN        NaN   
2        (37.82142504165362, -122.28732679694832)      OPEN        NaN   
3        (37.75306101818787, -122.16312297802386)      OPEN        NaN   
4          (37.785503590812, -122.20544605709021)      OPEN        NaN   
...                                           ...       ...        ...   
852536    (37.80086333056584, -122.2728949987278)      OPEN        NaN   
852537     (37.7990150403045, -122.2729668018932)      OPEN        NaN   
852538   (37.78429373521214, -122.19783353979535)      OPEN        NaN   
852539   (37.75920504225326, -122.18706682059627)  UNFUNDED        NaN   
852540  (37.781792932032374, -122.24281520941372)  UNFUNDED        NaN   

       DATETIMECLOSED          SRX          SRY COUNCILDISTRICT BEAT  \
0                 NaN  6057859.838  2114424.905            CCD2  19X   
1                 NaN  6061437.527  2114126.459            CCD5  21X   
2                 NaN  6045445.169  2126491.007            CCD3  07X   
3                 NaN  6080871.846  2100940.775            CCD7  35X   
4                 NaN  6068853.201  2112972.033            CCD4  24Y   
...               ...          ...          ...             ...  ...   
852536            NaN  6049471.080  2118926.060            CCD2  03X   
852537            NaN  6049437.589  2118253.569            CCD2  03X   
852538            NaN  6071044.715  2112491.573            CCD6  28X   
852539            NaN  6073990.955  2103301.471            CCD7  30X   
852540            NaN  6058030.915  2111820.075            CCD5  19X   

                          PROBADDRESS     City State  
0                       1438 16TH AVE  Oakland    CA  
1                      2324 E 21ST ST  Oakland    CA  
2                  ETTIE ST & 28TH ST  Oakland    CA  
3                    9227 HILLSIDE ST  Oakland    CA  
4                  3901 ALLENDALE AVE  Oakland    CA  
...                               ...      ...   ...  
852536                   988 BROADWAY  Oakland    CA  
852537           7TH ST & FRANKLIN ST  Oakland    CA  
852538             3216 COURTLAND AVE  Oakland    CA  
852539  73RD AVE & INTERNATIONAL BLVD  Oakland    CA  
852540               2000 EMBARCADERO  Oakland    CA  

[852541 rows x 16 columns]

### Using df.info()
Lastly, I can use df.info() to get additional key information on the dataframe. Perhaps most useful is the column names and what dtype each column is. 

### Refresher on dtypes
As a refresher of what we learned in class, 

> **dtype** stands for **data type** and comes from the `numpy` module, which was automatically imporated as part of the pandas module. dt.info() will display `int64` instead of `int`, and `float64` instead of `float`. We also learned in class that `object` is pretty close to a `string`, but not exactly the same, as `object` is a sequence of bytes (machine readable), while `string` is a sequence of characters.


In [12]:
oakland_311_original.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 852541 entries, 0 to 852540
Data columns (total 16 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   REQUESTID        852541 non-null  int64  
 1   DATETIMEINIT     852541 non-null  object 
 2   SOURCE           852541 non-null  object 
 3   DESCRIPTION      852541 non-null  object 
 4   REQCATEGORY      852460 non-null  object 
 5   REQADDRESS       836654 non-null  object 
 6   STATUS           852540 non-null  object 
 7   REFERREDTO       70168 non-null   object 
 8   DATETIMECLOSED   610875 non-null  object 
 9   SRX              836654 non-null  float64
 10  SRY              836654 non-null  float64
 11  COUNCILDISTRICT  809790 non-null  object 
 12  BEAT             812245 non-null  object 
 13  PROBADDRESS      852064 non-null  object 
 14  City             852541 non-null  object 
 15  State            852541 non-null  object 
dtypes: float64(2), int64(1), object(13)
me

## Checking to see if we need to convert any columns for analysis purposes
As was the case with the Berkeley 311 data, the Oakland 311 data has columns for the date that each case opened (DATETIMEINIT, column 1 and DATETIMECLOSED, column 8). Here I will use df.min() and df.max() to verify that these columns are not in fact sorted by actual date and instead by string. Attempting to execute this command will return wrong answers, sorted by string instead of actual date. This is because their Dtype is object. This means I'll need to convert these columns to the datetime format, which I'll do in the next section.

In [13]:
oakland_311_original['DATETIMEINIT'].min()

'01/01/2011 05:51:45 PM'

In [14]:
oakland_311_original['DATETIMEINIT'].max()

'12/31/2020 12:59:06 PM'

## Step 4: Properly type the Oakland 311 data
### Copy the original dataframe

Before I convert the 2 columns, I'm going to copy the original dataframe into a new dataframe using `df.copy()`. This way I can still have the original dataframe saved in case I run into any problems later. I'll also be able to compare the edited dataframe with the original dataframe to check for errors. I'll name the copy dataframe simply `oakland_311`.

In [43]:
oakland_311_copy = oakland_311_original.copy()

### Converting the date opened and closed columns to `datetime`

Now that I have the copy, I can convert the two columns from `object` dtype to `datetime` Dtype. I'll convert **DATETIMEINIT** first so I can compare the two columns. The syntax for this conversion is:

```python
df['column_name'] = pd.to_datetime(df['column_name'])
```

In [55]:
oakland_311_copy['DATETIMEINIT'] = pd.to_datetime(oakland_311_copy['DATETIMEINIT'])

As noted in class, `oakland_311['Date_Opened']` is called a pandas **series**. Now I can use df.head() and df.info() to compare the converted DATETIMEINIT column to the DATETIMECLOSED column that I have not yet converted:

In [56]:
oakland_311_copy.head()

REQUESTID        DATETIMEINIT       SOURCE  \
0    1145885 2021-09-09 18:08:56  SeeClickFix   
1    1146984 2021-09-14 08:08:50        Phone   
2    1150718 2021-09-27 07:52:03  SeeClickFix   
3    1151336 2021-09-28 15:37:51  SeeClickFix   
4    1151681 2021-09-29 16:18:33  SeeClickFix   

                                  DESCRIPTION  REQCATEGORY  \
0     Illegal Dumping (Enforcement Potential)  ENVIRON_ENF   
1     Illegal Dumping (Enforcement Potential)  ENVIRON_ENF   
2  Illegal Dumping - debris, appliances, etc.      ILLDUMP   
3                 Parking - Abandoned Vehicle       POLICE   
4                 Parking - Abandoned Vehicle       POLICE   

                                  REQADDRESS STATUS REFERREDTO DATETIMECLOSED  \
0  (37.788936670596335, -122.24357499716558)   OPEN        NaN            NaT   
1   (37.78829957727266, -122.23117705915959)   OPEN        NaN            NaT   
2   (37.82142504165362, -122.28732679694832)   OPEN        NaN            NaT   
3   (37.75306101818787, -122.16312297802386)   OPEN        NaN            NaT   
4     (37.785503590812, -122.20544605709021)   OPEN        NaN            NaT   

           SRX          SRY COUNCILDISTRICT BEAT         PROBADDRESS     City  \
0  6057859.838  2114424.905            CCD2  19X       1438 16TH AVE  Oakland   
1  6061437.527  2114126.459            CCD5  21X      2324 E 21ST ST  Oakland   
2  6045445.169  2126491.007            CCD3  07X  ETTIE ST & 28TH ST  Oakland   
3  6080871.846  2100940.775            CCD7  35X    9227 HILLSIDE ST  Oakland   
4  6068853.201  2112972.033            CCD4  24Y  3901 ALLENDALE AVE  Oakland   

  State  
0    CA  
1    CA  
2    CA  
3    CA  
4    CA

In [58]:
oakland_311_copy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 852541 entries, 0 to 852540
Data columns (total 16 columns):
 #   Column           Non-Null Count   Dtype         
---  ------           --------------   -----         
 0   REQUESTID        852541 non-null  int64         
 1   DATETIMEINIT     852541 non-null  datetime64[ns]
 2   SOURCE           852541 non-null  object        
 3   DESCRIPTION      852541 non-null  object        
 4   REQCATEGORY      852460 non-null  object        
 5   REQADDRESS       836654 non-null  object        
 6   STATUS           852540 non-null  object        
 7   REFERREDTO       70168 non-null   object        
 8   DATETIMECLOSED   610875 non-null  datetime64[ns]
 9   SRX              836654 non-null  float64       
 10  SRY              836654 non-null  float64       
 11  COUNCILDISTRICT  809790 non-null  object        
 12  BEAT             812245 non-null  object        
 13  PROBADDRESS      852064 non-null  object        
 14  City             852

As we can see above, the format of the DATETIMEINIT column has changed, as has its Dtype. It is now a datetime64 Dtype.

Now I can convert the DATETIMECLOSED row:

In [52]:
oakland_311_copy['DATETIMECLOSED'] = pd.to_datetime(oakland_311_copy['DATETIMECLOSED'])

I'll also take a look again to make sure the change has occurred:

In [59]:
oakland_311_copy.head()

REQUESTID        DATETIMEINIT       SOURCE  \
0    1145885 2021-09-09 18:08:56  SeeClickFix   
1    1146984 2021-09-14 08:08:50        Phone   
2    1150718 2021-09-27 07:52:03  SeeClickFix   
3    1151336 2021-09-28 15:37:51  SeeClickFix   
4    1151681 2021-09-29 16:18:33  SeeClickFix   

                                  DESCRIPTION  REQCATEGORY  \
0     Illegal Dumping (Enforcement Potential)  ENVIRON_ENF   
1     Illegal Dumping (Enforcement Potential)  ENVIRON_ENF   
2  Illegal Dumping - debris, appliances, etc.      ILLDUMP   
3                 Parking - Abandoned Vehicle       POLICE   
4                 Parking - Abandoned Vehicle       POLICE   

                                  REQADDRESS STATUS REFERREDTO DATETIMECLOSED  \
0  (37.788936670596335, -122.24357499716558)   OPEN        NaN            NaT   
1   (37.78829957727266, -122.23117705915959)   OPEN        NaN            NaT   
2   (37.82142504165362, -122.28732679694832)   OPEN        NaN            NaT   
3   (37.75306101818787, -122.16312297802386)   OPEN        NaN            NaT   
4     (37.785503590812, -122.20544605709021)   OPEN        NaN            NaT   

           SRX          SRY COUNCILDISTRICT BEAT         PROBADDRESS     City  \
0  6057859.838  2114424.905            CCD2  19X       1438 16TH AVE  Oakland   
1  6061437.527  2114126.459            CCD5  21X      2324 E 21ST ST  Oakland   
2  6045445.169  2126491.007            CCD3  07X  ETTIE ST & 28TH ST  Oakland   
3  6080871.846  2100940.775            CCD7  35X    9227 HILLSIDE ST  Oakland   
4  6068853.201  2112972.033            CCD4  24Y  3901 ALLENDALE AVE  Oakland   

  State  
0    CA  
1    CA  
2    CA  
3    CA  
4    CA

We can notice here (and indeed may have noticed before in the original oakland 311 file) that some number of the cells under the DATETIMECLOSED say 'NaN', which indicates that there is no data available. A quick glance across the table shows that this is in the case of cases that have a status of OPEN. Because of this missing data, we will not be able to find the difference between the DATETIMEINIT and DATETIMECLOSED columns for every case, but only for closed cases that have data for both the DATETIMEINIT and DATETIMECLOSED columns.

In [60]:
oakland_311_copy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 852541 entries, 0 to 852540
Data columns (total 16 columns):
 #   Column           Non-Null Count   Dtype         
---  ------           --------------   -----         
 0   REQUESTID        852541 non-null  int64         
 1   DATETIMEINIT     852541 non-null  datetime64[ns]
 2   SOURCE           852541 non-null  object        
 3   DESCRIPTION      852541 non-null  object        
 4   REQCATEGORY      852460 non-null  object        
 5   REQADDRESS       836654 non-null  object        
 6   STATUS           852540 non-null  object        
 7   REFERREDTO       70168 non-null   object        
 8   DATETIMECLOSED   610875 non-null  datetime64[ns]
 9   SRX              836654 non-null  float64       
 10  SRY              836654 non-null  float64       
 11  COUNCILDISTRICT  809790 non-null  object        
 12  BEAT             812245 non-null  object        
 13  PROBADDRESS      852064 non-null  object        
 14  City             852

### Using .min() and .max() to now find the correct earliest and latest entries
Now I can find the earliest date and the latest date of both columns, by performing `series.min()` and `series.max()` on these series.

In [61]:
oakland_311_copy['DATETIMEINIT'].min()

Timestamp('2009-07-01 08:05:36')

In [62]:
oakland_311_copy['DATETIMEINIT'].max()

Timestamp('2021-10-30 01:29:15')

In [63]:
oakland_311_copy['DATETIMECLOSED'].min()

Timestamp('2006-09-08 15:05:19')

In [64]:
oakland_311_copy['DATETIMECLOSED'].max()

Timestamp('2021-10-29 16:24:38')

The min and max tell us that the year 2009 is not complete (as of course is also true of 2021 because it's still happening). If I analyze the data by year later, I might consider excluding these years.

### Getting the difference of the two date columns

Using Pandas, I can find the difference of two dates by subtracting one datetime column from another. I'll now create a new column called **CLOSETIME** that shows how long it took for a case to be closed:

In [65]:
oakland_311_copy['CLOSETIME'] = oakland_311_copy['DATETIMECLOSED'] - oakland_311_copy['DATETIMEINIT']

The resulting column should not be a `datetime` dtype, but rather a `timedelta` dtype. As a reminder from class, the term "delta" is often used to mean "change." I'll now use df.head() again to make sure the new column has been added:

In [75]:
oakland_311_copy.head(40)

REQUESTID        DATETIMEINIT          SOURCE  \
0     1145885 2021-09-09 18:08:56     SeeClickFix   
1     1146984 2021-09-14 08:08:50           Phone   
2     1150718 2021-09-27 07:52:03     SeeClickFix   
3     1151336 2021-09-28 15:37:51     SeeClickFix   
4     1151681 2021-09-29 16:18:33     SeeClickFix   
5     1151692 2021-09-29 16:47:58           Other   
6     1151727 2021-09-29 20:10:51     SeeClickFix   
7     1147659 2021-09-15 19:46:41     SeeClickFix   
8     1149819 2021-09-23 10:09:38           Phone   
9     1149831 2021-09-23 10:31:17           Phone   
10    1151508 2021-09-29 09:42:45           Phone   
11    1151531 2021-09-29 10:44:12     SeeClickFix   
12    1151540 2021-09-29 11:01:47     SeeClickFix   
13    1151552 2021-09-29 11:25:58           Email   
14    1149804 2021-09-23 09:51:49           Phone   
15    1150199 2021-09-24 10:47:26     SeeClickFix   
16     369975 2011-05-12 14:43:40  Phone or Email   
17    1151561 2021-09-29 12:01:00     SeeClickFix   
18    1151564 2021-09-29 12:06:33     SeeClickFix   
19    1151569 2021-09-29 12:14:38     SeeClickFix   
20    1151580 2021-09-29 12:34:33     SeeClickFix   
21    1151588 2021-09-29 12:56:14     SeeClickFix   
22    1151623 2021-09-29 14:33:27           Phone   
23    1151634 2021-09-29 14:56:00           Phone   
24    1151639 2021-09-29 15:02:45     SeeClickFix   
25    1151665 2021-09-29 16:00:12           Phone   
26    1144273 2021-09-02 21:13:10     SeeClickFix   
27    1146047 2021-09-10 11:28:26           Email   
28    1151441 2021-09-29 06:45:44     SeeClickFix   
29    1152064 2021-10-01 03:37:20     SeeClickFix   
30    1142452 2021-08-27 13:03:44     SeeClickFix   
31    1151128 2021-09-28 09:41:28     SeeClickFix   
32    1152171 2021-10-01 11:01:30           Phone   
33    1152203 2021-10-01 12:04:32     SeeClickFix   
34    1152142 2021-10-01 10:18:24     SeeClickFix   
35    1152299 2021-10-01 15:59:17           Phone   
36    1152129 2021-10-01 09:54:19     SeeClickFix   
37    1152235 2021-10-01 13:56:58           Phone   
38    1152180 2021-10-01 11:14:32           Phone   
39    1152282 2021-10-01 15:36:02           Email   

                                   DESCRIPTION    REQCATEGORY  \
0      Illegal Dumping (Enforcement Potential)    ENVIRON_ENF   
1      Illegal Dumping (Enforcement Potential)    ENVIRON_ENF   
2   Illegal Dumping - debris, appliances, etc.        ILLDUMP   
3                  Parking - Abandoned Vehicle         POLICE   
4                  Parking - Abandoned Vehicle         POLICE   
5                  Parking - Abandoned Vehicle         POLICE   
6                  Parking - Abandoned Vehicle         POLICE   
7      Illegal Dumping (Enforcement Potential)    ENVIRON_ENF   
8   Illegal Dumping - debris, appliances, etc.        ILLDUMP   
9   Illegal Dumping - debris, appliances, etc.        ILLDUMP   
10                 Parking - Abandoned Vehicle         POLICE   
11                 Parking - Abandoned Vehicle         POLICE   
12                 Parking - Abandoned Vehicle         POLICE   
13                City Bldg - Structural Issue      BLDGMAINT   
14        Illegal Dumping – mattress/boxspring        ILLDUMP   
15  Illegal Dumping - debris, appliances, etc.        ILLDUMP   
16                           Sidewalk - Damage   ROW_STREETSW   
17                 Parking - Abandoned Vehicle         POLICE   
18                 Parking - Abandoned Vehicle         POLICE   
19                 Parking - Abandoned Vehicle         POLICE   
20                 Parking - Abandoned Vehicle         POLICE   
21                 Parking - Abandoned Vehicle         POLICE   
22                 Parking - Abandoned Vehicle         POLICE   
23                 Parking - Abandoned Vehicle         POLICE   
24                 Parking - Abandoned Vehicle         POLICE   
25                 Parking - Abandoned Vehicle         POLICE   
26     Illegal Dumping (Enforcement Potential)    ENVIRON_ENF   
27     Ill

I used df.head(40) above to see more rows of the data, as I wanted to make sure that the operation worked for the rows for which there is data available for both DATETIMEINIT and DATETIMECLOSED. We see that the new CLOSETIME column has been created and that the operation seems to have worked for rows with complete data, such as rows 7 and 27.

## Step 5: Conducting some basic data analysis
Now that I've converted the columns, I can do some interesting operations on them, just like we did with the Berkeley 311 dataset in class. 

### Mean
Here I'll the mean of a column by calling series.mean():


In [76]:
oakland_311_copy['CLOSETIME'].mean()

Timedelta('49 days 19:45:09.623400861')

We can see here that the average time the city of Oakland took to close a case was nearly 50 days. This is for the whole dataset, but only includes cases that have closed. We know that more than 200,000 cases are still open, so this finding may not be representative. 

### Median

In [77]:
oakland_311_copy['CLOSETIME'].median()

Timedelta('4 days 22:03:37')

It looks like the median time was around 5 days, which seems consistent with what we saw in the Berkeley 311 data as well.

### Minimum

In [78]:
oakland_311_copy['CLOSETIME'].min()

Timedelta('-2981 days +00:37:47')

It appears that the minimum amount of time it took to close a case is appearing as a negative number here, which is obviously a mistake that would need to be looked into.

### Maximum

In [79]:
oakland_311_copy['CLOSETIME'].max()

Timedelta('4293 days 21:59:22')

It appears that the maximum amount of time it took to close a case was more than 4,293 days, or more than 11 years! This may have also been an error.

### Sorting the data

Now I'm going to sort the data to look at the top 5 cases that took the longest time to resolve. Just like in calss, I'll use the `df.sort_values()` method.

Next, I don't want to see the entire dataframe, just the first 10 rows. So I'm going to call `df.head(10)`.

In [80]:
oakland_311_copy.sort_values(by=['CLOSETIME'], ascending=False).head(5)

REQUESTID        DATETIMEINIT          SOURCE  \
46677     323595 2009-12-09 14:30:28  Phone or Email   
55197     357227 2011-01-11 16:11:55         Website   
57468     356540 2011-01-05 21:40:17         Website   
83958     370528 2011-05-18 15:53:54  Phone or Email   
75121     370530 2011-05-18 15:59:39  Phone or Email   

                                DESCRIPTION   REQCATEGORY  \
46677           Street Medians (Landscaped)         PARKS   
55197                     Sidewalk - Damage  ROW_STREETSW   
57468  Streets/Sidewalks - Unsafe Condition  ROW_STREETSW   
83958          Park - Landscape Maintenance         PARKS   
75121          Park - Landscape Maintenance         PARKS   

                                     REQADDRESS  STATUS REFERREDTO  \
46677  (37.82206394742191, -122.20671960118034)  CLOSED        NaN   
55197  (37.73561875315405, -122.17677744021242)  CLOSED        NaN   
57468   (37.81062199910106, -122.2589020860214)  CLOSED        NaN   
83958  (37.75119515507376, -122.15561532358669)  CLOSED        NaN   
75121  (37.75119515507376, -122.15561532358669)  CLOSED        NaN   

           DATETIMECLOSED          SRX          SRY COUNCILDISTRICT BEAT  \
46677 2021-09-11 12:29:50  6068728.000  2126289.000            CCD4  22X   
55197 2021-08-05 09:29:00  6076811.090  2094661.167            CCD7  31Z   
57468 2021-07-14 09:57:37  6053579.984  2122402.509            CCD3  14X   
83958 2021-09-11 12:08:19  6083030.000  2100223.000            CCD7  35X   
75121 2021-09-11 12:08:47  6083030.000  2100223.000            CCD7  35X   

                     PROBADDRESS     City State          CLOSETIME  
46677  PARK BLVD & MOUNTAIN BLVD  Oakland    CA 4293 days 21:59:22  
55197            10306 PIPPIN ST  Oakland    CA 3858 days 17:17:05  
57468          325 PARK VIEW TER  Oakland    CA 3842 days 12:17:20  
83958            FIRE STATION 26  Oakland    CA 3768 days 20:14:25  
75121            FIRE STATION 26  Oakland    CA 3768 days 20:09:08

These 5 cases all took thousands of days to resolve. This could be an error...or the beginning of an interesting story idea. Further investigation would be needed to determine which.

## Step 6: Cleaning the Data by Checking and Vetting it
### Checking to see if there is a unique identifier for every row.

Just like in class with the Berkeley 311 dataset, I will check to see if `REQUESTID` has a unique ID for every row, which is important for running most anlaysis.

I will get a count of the unique values by calling `.nunique()` on that column.

In [81]:
oakland_311_copy['REQUESTID'].nunique()

852541

Now I can compare that number with the total number of rows to see if they match. I'll call `len(df)` to find out, which we learned to use earlier to find the length of strings and datastructures:

In [82]:
len(oakland_311_copy)

852541

Unlike in the Berkeley 311 dataset, this Oakland 311 data seems to have the a request ID for every row, which is good news! This means we won't need to create new request IDs. Just for practice though, I'll still check to see if there are any duplicates. 

In [83]:
oakland_311_copy[oakland_311.duplicated()]

Empty DataFrame
Columns: [REQUESTID, DATETIMEINIT, SOURCE, DESCRIPTION, REQCATEGORY, REQADDRESS, STATUS, REFERREDTO, DATETIMECLOSED, SRX, SRY, COUNCILDISTRICT, BEAT, PROBADDRESS, City, State, CLOSETIME]
Index: []

As expected, there are no exact duplicates. Now we'll double check to see if there are any duplicates in the REQUEST ID column:

In [84]:
oakland_311_copy[oakland_311['REQUESTID'].duplicated()]

Empty DataFrame
Columns: [REQUESTID, DATETIMEINIT, SOURCE, DESCRIPTION, REQCATEGORY, REQADDRESS, STATUS, REFERREDTO, DATETIMECLOSED, SRX, SRY, COUNCILDISTRICT, BEAT, PROBADDRESS, City, State, CLOSETIME]
Index: []

Also as expected, there are no duplicates in the REQUESTID column either. Thus, there is no need to create a list of duplicates in the REQUESTID column by subsetting, since that list would be empty.

### How many `NaN`s do we have?
As mentioned above, this dataset has a lot of `NaN`s in the `DATETIMEINIT` and `DATETIMECLOSED` columns, and now in the new `CLOSETIME` column we created. Let's use df.info() to find out how many:

In [85]:
oakland_311_copy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 852541 entries, 0 to 852540
Data columns (total 17 columns):
 #   Column           Non-Null Count   Dtype          
---  ------           --------------   -----          
 0   REQUESTID        852541 non-null  int64          
 1   DATETIMEINIT     852541 non-null  datetime64[ns] 
 2   SOURCE           852541 non-null  object         
 3   DESCRIPTION      852541 non-null  object         
 4   REQCATEGORY      852460 non-null  object         
 5   REQADDRESS       836654 non-null  object         
 6   STATUS           852540 non-null  object         
 7   REFERREDTO       70168 non-null   object         
 8   DATETIMECLOSED   610875 non-null  datetime64[ns] 
 9   SRX              836654 non-null  float64        
 10  SRY              836654 non-null  float64        
 11  COUNCILDISTRICT  809790 non-null  object         
 12  BEAT             812245 non-null  object         
 13  PROBADDRESS      852064 non-null  object         
 14  City

According to this, it looks like the `DATETIMECLOSE` column is the one that has the most `NaN`s. We could no doubt subset the cases that are missing a `DATETIMECLOSE` date and run some analyses to find out if, for example, most of those cases were opened recently. 

### Assert

We learned in class that the keyword `assert` is a way to check if the length of the dataframe now matches the number of unique IDs. Although I don't really need to do this, I will do it anyway here for practice. 

In [86]:
assert len(oakland_311_copy) == oakland_311['REQUESTID'].nunique()

## Step 7: Exporting a clean version of the data to a csv
I will now use the `df.to_csv()` method to export a clean copy of the csv, which will allow me to instantly import the clean data in a new notebook instead of rerunning the code in this notebook. This is still valuable for me to do, since I did make some important changes, namely changing the dType for the columns 2 and 7 and adding a new column.

In [87]:
oakland_311_copy.to_csv('oakland_311_clean.csv', index=False)

Lastly, I'll try removing index=False to see what happens: 

In [3]:
# oakland_311_copy.to_csv('oakland_311_clean.csv')